lib/src/query_component.dart

```diff
@@ -152,4 +152,36 @@ class QueryComponent implements OnInit {
             InsertionOrder_Budget_BudgetSegment())
         .toList();
   }
+
+  DateTime _getMinStartDate(
+          List<InsertionOrder> insertionOrders) =>
+      insertionOrders
+          .map((io) => Util.convertProtoDateToDateTime(
+              io.budget.activeBudgetSegment.dateRange.startDate))
+          .reduce((date, minDate) => date.isBefore(minDate) ? date : minDate);
+
+  void _addSpentToInsertionOrders(List<InsertionOrder> insertionOrders,
+      Map<String, List<InsertionOrderDailySpend>> spendingMap) {
+    for (final io in insertionOrders) {
+      // Continue if the spending map doesn't have spend data for this io.
+      if (!spendingMap.containsKey(io.insertionOrderId)) continue;
+
+      final budgetUnit = io.budget.budgetUnit;
+      final flightStart = Util.convertProtoDateToDateTime(
+          io.budget.activeBudgetSegment.dateRange.startDate);
+      final flightEnd = Util.convertProtoDateToDateTime(
+          io.budget.activeBudgetSegment.dateRange.endDate);
+
+      final spent = spendingMap[io.insertionOrderId]
+          .where((dailySpend) =>
+              Util.isBetweenDates(dailySpend.date, flightStart, flightEnd))
+          .map((dailySpend) => budgetUnit ==
+                  InsertionOrder_Budget_BudgetUnit.BUDGET_UNIT_CURRENCY
+              ? dailySpend.revenue
+              : dailySpend.impression)
+          .reduce((spend, sum) => Util.addStringRevenue(spend, sum));
```

Comment:
this way you will end up iterating over the data twice (once to build the map and once to see which values should be added). Imo i think this is the right way to do it because the division of responsibility and the code readability is more important.